circleguard/gui.py

```diff
@@ -279,71 +305,416 @@ def __init__(self):
         self.terminal.append(message)
 
-class MainWindow(QWidget):
+class MainWindow(QFrame):
     def __init__(self):
         super().__init__()
 
-        self.tab_widget = QTabWidget()
+        self.tabs = QTabWidget()
         self.main_tab = MainTab()
         self.results_tab = ResultsTab()
         self.queue_tab = QueueTab()
         self.thresholds_tab = ThresholdsTab()
         self.settings_tab = SettingsTab()
-        self.tab_widget.addTab(self.main_tab, "Main")
-        self.tab_widget.addTab(self.results_tab, "Results")
-        self.tab_widget.addTab(self.queue_tab, "Queue")
-        self.tab_widget.addTab(self.thresholds_tab, "Thresholds")
-        self.tab_widget.addTab(self.settings_tab, "Settings")
+        self.tabs.addTab(self.main_tab, "Main")
+        self.tabs.addTab(self.results_tab, "Results")
+        self.tabs.addTab(self.queue_tab, "Queue")
+        self.tabs.addTab(self.thresholds_tab, "Thresholds")
+        self.tabs.addTab(self.settings_tab, "Settings")
         # so when we switch from settings tab to main tab, whatever tab we're on gets changed if we delete our api key
-        self.tab_widget.currentChanged.connect(self.main_tab.switch_run_button)
+        self.tabs.currentChanged.connect(self.main_tab.check_run_button)
 
         self.layout = QVBoxLayout()
-        self.layout.addWidget(self.tab_widget)
+        self.layout.addWidget(self.tabs)
         self.layout.setContentsMargins(10, 10, 10, 0)
         self.setLayout(self.layout)
 
 
-class MainTab(QWidget):
-    set_progressbar_signal = pyqtSignal(int)  # max progress
-    increment_progressbar_signal = pyqtSignal(int)  # increment value
+class ScrollableLoadablesWidget(QFrame):
+    def __init__(self):
+        super().__init__()
+        self.layout = QVBoxLayout()
+        self.layout.setAlignment(Qt.AlignTop)
+        self.setLayout(self.layout)
+
+
+class ScrollableChecksWidget(QFrame):
+    def __init__(self):
+        super().__init__()
+        self.layout = QVBoxLayout()
+        self.layout.setAlignment(Qt.AlignTop)
+        self.setLayout(self.layout)
+
+
+class DropArea(QFrame):
+    # style largely taken from
+    # https://doc.qt.io/qt-5/qtwidgets-draganddrop-dropsite-example.html
+    # (we use a QFrame instead of a QLabel so we can have a layout and
+    # add new items to it)
+    def __init__(self):
+        super().__init__()
+
+        self.loadable_ids = [] # ids of loadables already in this drop area
+        self.loadables = [] # LoadableInChecks in this DropArea
+        self.setMinimumSize(0, 100)
+        self.setFrameStyle(QFrame.Sunken | QFrame.StyledPanel)
+        self.setAcceptDrops(True)
+
+        self.layout = QVBoxLayout()
+        self.layout.setAlignment(Qt.AlignTop)
+        self.setLayout(self.layout)
+
+    def dragEnterEvent(self, event):
+        # need to accept the event so qt gives us the DropEvent
+        # https://doc.qt.io/qt-5/qdropevent.html#details
+        event.acceptProposedAction()
```

Comment:
Should probably check that there is `circleguard/` data in the mimedata; atm you can drop anything in a Check and it will try to parse that.
![](https://i.imgur.com/gNbc9uv.png)

Reply:
Good point.

took the liberty of rewriting our handling of mime type as well. `application/` type is for data handled by applications and `x` is for "non-standard" mime types. Don't think we were using the MIME standard correctly before.